src/mapport.cpp

```diff
@@ -0,0 +1,320 @@
+                        AddLocal(resolved, LOCAL_MAPPED);
+                    }
+                } else {
+                    LogPrintf("UPnP: GetExternalIPAddress failed.\n");
+                }
+            }
+        }
+
+        std::string strDesc = PACKAGE_NAME " " + FormatFullVersion();
+
+        do {
+            r = UPNP_AddPortMapping(urls.controlURL, data.first.servicetype, port.c_str(), port.c_str(), lanaddr, strDesc.c_str(), "TCP", 0, "0");
+
+            if (r != UPNPCOMMAND_SUCCESS) {
+                ret = false;
+                LogPrintf("AddPortMapping(%s, %s, %s) failed with code %d (%s)\n", port, port, lanaddr, r, strupnperror(r));
+                break;
+            } else {
+                ret = true;
+                g_mapport_current_proto = MapPortProto::UPNP;
+                LogPrintf("UPnP Port Mapping successful.\n");
+            }
+        } while (g_mapport_interrupt.sleep_for(PORT_MAPPING_REANNOUNCE_PERIOD));
+
+        r = UPNP_DeletePortMapping(urls.controlURL, data.first.servicetype, port.c_str(), "TCP", 0);
+        LogPrintf("UPNP_DeletePortMapping() returned: %d\n", r);
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        FreeUPNPUrls(&urls);
+    } else {
+        LogPrintf("No valid UPnP IGDs found\n");
+        freeUPNPDevlist(devlist); devlist = nullptr;
+        if (r != 0)
+            FreeUPNPUrls(&urls);
+    }
+
+    return ret;
+}
+#endif // #ifdef USE_UPNP
+
+static void ThreadMapPort()
+{
+    bool ok;
+    do {
+        ok = false;
+
+#ifdef USE_NATPMP
+        if (g_mapport_target_proto & MapPortProto::NAT_PMP) {
+            ok = ThreadNatpmp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_NATPMP
+
+#ifdef USE_UPNP
+        if (g_mapport_target_proto & MapPortProto::UPNP) {
+            ok = ThreadUpnp();
+            if (ok) continue;
+        }
+#endif // #ifdef USE_UPNP
+
+        if (g_mapport_target_proto == MapPortProto::NONE) {
+            g_mapport_current_proto = MapPortProto::NONE;
+            return;
+        }
+
+    } while (ok || g_mapport_interrupt.sleep_for(PORT_MAPPING_RETRY_PERIOD));
+}
+
+void StartMapPort(bool use_natpmp, bool use_upnp)
+{
+    if (use_natpmp) {
+        g_mapport_target_proto |= MapPortProto::NAT_PMP;
+    } else {
+        g_mapport_target_proto &= ~MapPortProto::NAT_PMP;
+    }
+
+    if (use_upnp) {
+        g_mapport_target_proto |= MapPortProto::UPNP;
+    } else {
+        g_mapport_target_proto &= ~MapPortProto::UPNP;
+    }
+
+    if (g_mapport_current_proto == MapPortProto::NONE) {
+        if (g_mapport_target_proto == MapPortProto::NONE) {
+            return;
+        }
+    } else if (g_mapport_target_proto & g_mapport_current_proto) {
+        return;
+    }
+
+    if (g_mapport_target_proto == MapPortProto::NONE) {
+        InterruptMapPort();
+        StopMapPort();
+        return;
+    }
+
+    if (g_mapport_thread.joinable()) {
+        g_mapport_interrupt();
```

Comment:
Won't t his cause the thread to shut down?

Reply:
Fixed in the latest [push](https://github.com/bitcoin/bitcoin/pull/18077#issuecomment-589405068).